File: 20221001_Saturday/20221001.go
// Leetcode: 91. Decode Ways.
// https://leetcode.com/problems/decode-ways/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 6 ms, faster than 12.16% of Go online submissions for Decode Ways.
// Memory Usage: 2 MB, less than 45.59% of Go online submissions for Decode Ways.
import "strconv"
func numDecodings(s string) int {
  dp := make([]int, len(s) + 1)
  dp[0] = 1 // One way to reach the first digit.
  for i :=0 ; i < len(s) ; i += 1{
    if '0' == s[i] {continue}
    if i < (len(s)-1){
      // https://stackoverflow.com/a/29841190/8574922
      num, _ := strconv.ParseInt(s[i:i+2], 10, 32) 
      if num >= 10 && num <= 26{
        dp[i+2] += dp[i]
      }
    }
    dp[i+1] += dp[i]
  }
  return dp[len(s)]
}

File: 20221001_Saturday/20221001.rb
# Leetcode: 91. Decode Ways.
# https://leetcode.com/problems/decode-ways/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 179 ms, faster than 11.32% of Ruby online submissions for Decode Ways.
# Memory Usage: 211.1 MB, less than 45.28% of Ruby online submissions for Decode Ways.
# @param {String} s
# @return {Integer}
def num_decodings(s)
  dp = Array.new(s.size+1,0)
  dp[0] = 1
  s.chars.each_with_index do |c,i|
    digit = c.to_i
    if ?0 == c
      next
    elsif digit.between?(3,9)
      dp[i+1] += dp[i]
    elsif i < (s.size - 1) && s[i..i+1].to_i.between?(10,26)
      dp[i+1] += dp[i]
      dp[i+2] += dp[i]
    else
      dp[i+1] += dp[i]
    end
  end
  dp.last
end
